Add cluster-wide setting for total shard limit

dakrone opened the pull request:
This adds the `cluster.routing.allocation.total_shards_per_node`
setting, which limits the total number of shards across all indices on
each node. It defaults to -1 and can be dynamically configured.

Resolves #14456

jpountz commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
 
     public static final String NAME = "shards_limit";
 
+    private volatile Integer clusterShardLimit;
```
Looking at the code, it seems like it can never be null, so should it be an int instead?

dakrone replied:
Definitely, will change this.

jpountz commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
-        if (totalShardsPerNode <= 0) {
-            return allocation.decision(Decision.YES, NAME, "total shard limit disabled: [%d] <= 0", totalShardsPerNode);
+    private Decision checkLimits(int indexShardLimit, @Nullable ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (indexShardLimit <= 0 && clusterShardLimit <= 0) {
```
can you add something like `final int clusterShardLimit = this.clusterShardLimit` to make sure the method is consistently using the same value of the limit even if there are concurrent updates of the setting?

dakrone replied:
Yep, good idea.

jpountz commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
+                if (!nodeShard.index().equals(shardRouting.index())) {
+                    continue;
+                }
+                nodeCount++;
```
I think the following would read better to me:

```java
if (nodeShard.index().equals(shardRouting.index())) {
    nodeCount++;
}
```

dakrone replied:
Yes it would, I'll change this.

jpountz commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
             }
-            nodeCount++;
+            totalCount++;
+            if (shardRouting != null) {
```
can ShardRouting really be null? the previous version of the class did not seem to care?

jpountz replied:
oh nevermind, I just found the method that called it with null :)

dakrone replied:
I'm changing this because the null is confusing anyway.

jpountz commented:
I left some comments but this part of the code is out of my comfort zone so it would be nice that someone else look at it. Instead of having -1 as a default, I'm wondering if we should have a crazy high default value instead (say 10000?) that might already protect users, for instance if they are not aware of the cost of shards yet or if a buggy script starts creating indices in a loop because of the auto-create index feature? (I don't have a strong opinion here, just raising the idea for discussion)

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -24,13 +24,16 @@
 
 /**
  * This {@link AllocationDecider} limits the number of shards per node on a per
  * index basis. The allocator prevents a single node to hold more than
```
Here it still says `on a per index basis` -> should be corrected.

dakrone replied:
Will fix this.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
-            return allocation.decision(Decision.YES, NAME, "total shard limit disabled: [%d] <= 0", totalShardsPerNode);
-        }
+        // Here the total is incremented by 1 because it's a "remain" check instead of an allocation check
+        return checkLimits(indexMd.getSettings().getAsInt(INDEX_TOTAL_SHARDS_PER_NODE, -2) + 1, shardRouting, node, allocation);
```
What if the user specifies 0 here? Previously that meant unbounded.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -50,66 +53,88 @@
         }
-        return allocation.decision(Decision.YES, NAME, "shard count under limit [%d] of total shards per node", totalShardsPerNode);
+        return allocation.decision(Decision.YES, NAME, "shard count under index limit [%d] and node limit [%d] of total shards per node",
+                indexShardLimit, clusterShardLimit);
```
with inflating `indexShardLimit` by 1 in `canRemain`, this message might be confusing.

dakrone replied:
I changed the way it works so it is less confusing now.

ywelsch commented:
Added a few comments. I would not to inflate the indexShardLimit number in `canRemain`, but add a boolean parameter to `checkLimits`. This keeps decision messages sane but needs a distinction on the last if block.

dakrone commented:
I pushed a change addressing the feedback. Additionally, I changed the single method out into three distinct methods because, as @ywelsch mentioned, having a single one could lead to confusing messages as far as numbers are concerned.

Thanks for feedback @jpountz and @ywelsch

jpountz commented:
LGTM from my end

ywelsch commented:
LGTM 2. One last nitpick (but feel free to push): I was confused a bit by the variable names `nodeCount` and `totalCount`, maybe they could be renamed to sth like `indexShardCount` and to `nodeShardCount` ?

jpountz commented:
+1 on the variable name change
